Added ut mock function.

emccorkle opened the pull request:
## Description
This adds `ut mock function`, `ut expect call`, and `ut verify mock` that are useful when testing callbacks.

```
Button Box Test = ut test case("Button Box")
    <<Setup(Expr(
        mock = ut mock function({this}, Expr(1+1))
    ))
    <<Teardown(Expr(
        ut verify mock(mock)
    ));
	
ut test(Button Box Test, "Calls Callback", Expr(
    ut expect call(mock, {ut enabled(1)});
    New Window("", btn = Button Box("Hi", << Set Function(Function({this}, mock:fn(this)))));
    btn << Click;
));
```

## Checklist

- [x] **I am adding new or changing current functionality.**
  - [x] I have added or updated the tests for the new or changed functionality in `Tests/UnitTests`.
  - [x] I have added note(s) to `CHANGELOG.md` as necessary.

## Contributing

- [x] I have read and agree to the [Contributor Agreement](https://github.com/sassoftware/jsl-hamcrest/blob/master/ContributorAgreement.txt)

<!--- Replace your name and e-mail below -->
Signed-off-by: Evan McCorkle <Evan.McCorkle@jmp.com>

emccorkle committed:
Added ut mock function.

JustinChilton commented on Source/Core/Mock.jsl:
```diff
@@ -0,0 +1,225 @@
+		// Notice we are putting the parameter names in here, when ut verify mock call is actually called
+		// we will be inside the function call and these will resolve to locals with the actual argument values.
+		ut verify mock call(Expr(obj), Expr(fnarg1)); // Prepend args onto given function body
+	), 1); // First thing as soon as the function is called
```
What about this instead? Have not confirmed, but this looks like what we created the `ut merge expressions` function for, to merge them nicely with just one `Glue()`.
```
ut merge expressions(EvalExpr({ut verify mock call(Expr(obj), Expr(fnarg1)), body}));
```

JustinChilton commented on Source/Core/Mock.jsl:
```diff
@@ -0,0 +1,225 @@
+		));
+		---------
+*/
+ut mock function = Expr(Function({fnarg1, fnarg2, fnarg3="NOT PROVIDED"}, {obj, body},
```
Not saying this is bad, but all of our other documented functions put the full definition in the call to `ut define documented function`.  The expr would get overwritten in the function call later on.

emccorkle committed:
Address review comments.

- Using `ut merge expressions` instead of working around Glue()
- Test for multi-part mock function body with Glue()
- Moved function declaration into `ut define documented function`
